blueman/gui/applet/PluginDialog.py

```diff
@@ -159,45 +169,97 @@ def __init__(self, applet: "BluemanApplet") -> None:
-        b = self.list.get(iter2, "activatable", "name")
+    def _add_plugin_action(self, name: str, state: bool, activatable: bool) -> None:
+        logging.debug(f"adding action: {name}")
+        action = Gio.SimpleAction.new_stateful(
+            name, None, GLib.Variant.new_boolean(False)
+        )
+        action.set_property("enabled", activatable)
+        action.set_state(GLib.Variant.new_boolean(state))
+        self.add_action(action)
+        action.connect("change-state", self._on_plugin_toggle)
+
+    def _widget_factory(self, item: GObject.Object, _data: Optional[object] = None) -> Gtk.Widget:
+        assert isinstance(item, PluginItem)
+        box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=5, visible=True)
+
+        checkbutton = Gtk.CheckButton(visible=True, sensitive=item.props.activatable)
+        box.add(checkbutton)
+        checkbutton.set_action_name(f"win.{item.props.plugin_name}")
+
+        self._add_plugin_action(item.props.plugin_name, item.props.enabled, item.props.activatable)
+        # Set active after adding action
+        checkbutton.set_active(item.props.enabled)
 
-        if (a["activatable"] and b["activatable"]) or (not a["activatable"] and not b["activatable"]):
-            if a["name"] == b["name"]:
+        plugin_im = Gtk.Image(icon_name=item.props.icon_name, visible=True)
+        box.add(plugin_im)
+
+        label = Gtk.Label(label=item.props.description, use_markup=True, visible=True)
+        box.add(label)
+        return box
+
+    def _model_sort_func(self, item1: Optional[object], item2: Optional[object], _data: Optional[object] = None) -> int:
+        assert isinstance(item1, PluginItem)
+        assert isinstance(item2, PluginItem)
+
+        if (item1.props.activatable and item2.props.activatable) or (not item1.props.activatable and
+                                                                     not item2.props.activatable):
+            if item1.props.plugin_name == item2.props.plugin_name:
                 return 0
-            if a["name"] < b["name"]:
+            if item1.props.plugin_name < item2.props.plugin_name:
                 return -1
-            else:
-                return 1
-
+            return 1
         else:
-            if a["activatable"] and not b["activatable"]:
+            if item1.props.activatable and not item2.props.activatable:
                 return -1
-            elif not a["activatable"] and b["activatable"]:
+            if not item1.props.activatable and item2.props.activatable:
                 return 1
-            else:
-                return 0
+            return 0
 
-    def _on_close(self, _widget: Gtk.Widget, _event: Gdk.Event) -> bool:
-        self.applet.Plugins.disconnect(self.sig_a)
-        self.applet.Plugins.disconnect(self.sig_b)
-        return False
+    def _on_plugin_toggle(self, action: Gio.SimpleAction, state: GLib.Variant) -> None:
+        action.set_state(state)
```

Comment:
Not sure if this should not better be deferred until after the below branches, so that in case `_ask_unload` gets called and returns `False`-ish the state does not get changed here and then back there.

Reply:
I ended up with widget and action out of sync in certain situations.

For example, let's say the user wants to enable the `PPPSupport` which conflicts with the `NMDUNSupport`. In the callback the user decided not to enable plugin. The action is  `False` and setting it to `False` again does not send a signal to the widget to update it's state as there is no change. It's most likely the reason I messed up my GSettings.